allennlp/training/trainer.py

```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+                    self._writer.add_train_scalar("parameter_std/" + name, param.data.std(), batch_num_total)
+                    if param.grad is not None:
+                        self._writer.add_train_scalar("gradient_mean/" + name,
+                                                      param.grad.data.mean(),
+                                                      batch_num_total)
+                        self._writer.add_train_scalar("gradient_std/" + name,
+                                                      param.grad.data.std(),
+                                                      batch_num_total)
+                self._writer.add_train_scalar("loss/loss_train", metrics["loss"], batch_num_total)
+
+            # Log progress in no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
+                logger.info("Batch %d/%d: %s", batch_num, num_training_batches, description)
+                self._last_log = time.time()
+
+        # Return the metrics (and reset them)
+        return self._get_metrics_for_batch(train_loss, batch_num, reset=True)
+
+    def _next_epoch_validation(self, epoch: int, metrics: dict,
+                               validation_metric_per_epoch: List[float]) -> bool:
+        """
+        Validate one epoch, return True if we should stop early, False otherwise.
+        """
+        logger.info("Validating")
+        num_validation_batches = self._iterator.get_num_batches(self._validation_dataset)
+
+        # Switch to evaluation mode.
+        self._model.eval()
+
+        # Get tqdm for the validation batches
+        val_generator = self._iterator(self._validation_dataset, num_epochs=1)
+        val_generator_tqdm = tqdm.tqdm(val_generator,
+                                       disable=self._no_tqdm,
+                                       total=num_validation_batches)
+        batch_num = 0
+        val_loss = 0
+        for batch in val_generator_tqdm:
+            batch_num += 1
+
+            # Compute loss but don't backpropagate
+            loss = self._compute_loss(batch, for_training=False)
+            val_loss += loss.data.cpu().numpy()
+
+            # Update the description with the latest metrics
+            val_metrics = self._get_metrics_for_batch(val_loss, batch_num)
+            description = self._description_from_metrics(val_metrics)
+            val_generator_tqdm.set_description(description)
+
+            # Log progress in the no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
+                logger.info("Batch %d/%d: %s", batch_num, num_validation_batches, description)
+                self._last_log = time.time()
+
+        # Log parameter values to Tensorboard
+        val_metrics = self._get_metrics_for_batch(val_loss, batch_num, reset=True)
+        message_template = "Training %s : %3f    Validation %s : %3f "
+        for name, value in metrics.items():
```

Comment:
I'd _probably_ be fine with just having the `_train_epoch` function log the train metrics, and not have to pass them to this function.  If we do pass them here, I'd definitely at least call them `train_metrics` instead of just `metrics`.

Reply:
I changed the name